### PR TITLE
Timed chess

### LINKED ISSUE #3: Timed Chess
# Timed Chess

## Background
The king is getting old and is beginning to worry about when he is going to die. As such the king turned to you, the great sorcerer of time, for answers. As a master of the dark arts you knew exactly what you needed to do: create a countdown timer! 

## TODO
- [ ] Give each player a given amount of time to complete all their moves within.
- [ ] If a player runs out of time they should lose.
- [ ] When a player finishes a move they should be given an additional amount of time to their clock.

### PR DESCRIPTION
**Pull request by Sourc**
Fixes #3. 
Timed chess added. Some identention changes because of the IDE but it's still approved by the Google Styleguides. Also renamed the Layout enum to better match the added functionality.

### CODE REVIEW

File: src/main/java/chess/Board.java [hunk 1]
```diff
@@ -96,10 +99,9 @@ public boolean place(Piece piece, Position position) {
...
-   * @param newPosition   The position to chess.piece.move the chess.piece to.
-   * @return OK if move was successful.
-   *         INVALID_MOVE if the move was illegal.
-   *         PROMOTION_REQUIRED if a chess.piece needs promotion after the move.
+   * @param newPosition The position to chess.piece.move the chess.piece to.
+   * @return OK if move was successful. INVALID_MOVE if the move was illegal. PROMOTION_REQUIRED if
+   *     a chess.piece needs promotion after the move.
    */
   public MoveResult tryMove(Position piecePosition, Position newPosition) {
     if (!this.needsPromotion() && isValidMove(piecePosition, newPosition)) {
```
Comment (nolanderc) [inline src/main/java/chess/Board.java:109]: Currently the code does not check if the player is out of time/the game is over. What do you think about adding an additional check?

File: src/main/java/chess/Board.java [hunk 2]
```diff
@@ -209,6 +212,10 @@ public String toString() {
    * @return The type of check
    */
   public CheckType getCheck(Piece.Color color) {
+    if (timedChessController.hasLost(color)) {
+      return CheckType.CHECKMATE;
```
Comment (nolanderc) [inline src/main/java/chess/Board.java]: Should a new `CheckType` be added to clarify what happened? Something along the lines of `OUT_OF_TIME` perhaps?

File: src/main/java/chess/Board.java [hunk 3]
```diff
@@ -302,7 +310,7 @@ private boolean isColorInCheck(Piece.Color color) {
     return destinations;
   }
 
-  private void applyLayout(Layout layout) {
+  private void applyLayout(GameMode gameMode) {
```
Comment (nolanderc) [inline src/main/java/chess/Board.java]: Renaming `Layout` to `GameMode` was a great change. Renaming `applyLayout` to `applyGameMode` wouid reflect it's new purpose more clearly.

File: src/main/java/chess/TimeController.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+  private Timer startTimer;
+
+  /**
+   * Creates a time controller that will keep track of much time each player got left to play with.
+   *
+   * @param totalTime The total amount of time each player begins with, in minutes.
+   * @param timeIncreasePerMove The time given to the player when a move is made, in seconds.
+   */
+  TimeController(double totalTime, double timeIncreasePerMove) {
```
Comment (nolanderc) [inline src/main/java/chess/TimeController.java]: Is it possible that `totalTime` being in minutes and `timeIncreasePerMove` in seconds will confuse people having to use this constructor? Having them represent the same unit of time might be simpler.

File: src/main/java/chess/TimeController.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    this.totalTime = (long) (totalTime * 60000);
+    this.timeIncreasePerMove = (long) (timeIncreasePerMove * 1000);
+    active = true;
+  }
+
+  TimeController() {
+  }
+
+  void beginIn(long delay) {
```
Comment (nolanderc) [inline src/main/java/chess/TimeController.java]: What unit of time does `delay` have?

**Review by nolanderc — CHANGES_REQUESTED** [review]
These changes are looking good to me!
However, there are some minor changes which need to be addressed before a merge.

File: src/main/java/chess/Board.java (hunk 1 above)
Comment (Sourc) [reply src/main/java/chess/Board.java:109]: It does put the game into checkmate. I assumed that would be handled by the engine and did not look into it further.

**Review by nolanderc — APPROVED** [review]
Nicely done! I'm approving these changes.

**Pull request merged**